x/emissions/migrations/v3/migrate.go

```diff
@@ -86,11 +93,184 @@ func MigrateParams(store storetypes.KVStore, cdc codec.BinaryCodec) error {
+		err = feeRevenue.Unmarshal(topicFeeRevStore.Get(idArray))
+		if err != nil {
+			continue
+		}
+		var stake = cosmosMath.NewInt(0)
+		err = stake.Unmarshal(topicStakeStore.Get(idArray))
+		if err != nil {
+			continue
+		}
+		var previousWeight = alloraMath.NewDecFromInt64(0)
+		err = previousWeight.Unmarshal(topicPreviousWeightStore.Get(idArray))
+		if err != nil {
+			continue
+		}
+		// Get topic's latest weight
+		weight, err := getTopicWeight(
+			feeRevenue,
+			stake,
+			previousWeight,
+			oldMsg.EpochLength,
+			params.TopicRewardAlpha,
+			params.TopicRewardStakeImportance,
+			params.TopicRewardFeeRevenueImportance,
+			emissionsKeeper,
+		)
+		if err != nil {
+			continue
+		}
+		topicWeightData[oldMsg.Id] = weight
+		blockHeight := oldMsg.EpochLastEnded + oldMsg.EpochLength
+		sdkCtx.Logger().Warn(fmt.Sprintf("update blockHeight %d", blockHeight))
+		// If the weight less than minimum weight then skip this topic
+		if weight.Lt(params.MinTopicWeight) {
+			continue
+		}
+
+		cuLowestWeight := lowestWeight[blockHeight]
+		// Update lowest weight of topic per block
+		if cuLowestWeight.Weight.Equal(alloraMath.ZeroDec()) ||
+			weight.Lt(lowestWeight[blockHeight].Weight) {
+			cuLowestWeight = types.TopicIdWeightPair{
+				Weight:  weight,
+				TopicId: oldMsg.Id,
+			}
+		}
+
+		churningBlock[oldMsg.Id] = blockHeight
+
+		activeTopicIds := blockToActiveTopics[blockHeight]
+		activeTopicIds.TopicIds = append(activeTopicIds.TopicIds, oldMsg.Id)
+
+		// If number of active topic is over global param then remove lowest topic
+		if uint64(len(blockToActiveTopics[blockHeight].TopicIds)) > params.MaxActiveTopicsPerBlock {
+			// Remove from topicToNextPossibleChurningBlock
+			delete(churningBlock, lowestWeight[blockHeight].TopicId)
+			newActiveTopicIds := []types.TopicId{}
+			for i, id := range blockToActiveTopics[blockHeight].TopicIds {
+				if id == lowestWeight[blockHeight].TopicId {
+					newActiveTopicIds = append(blockToActiveTopics[blockHeight].TopicIds[:i],
+						blockToActiveTopics[blockHeight].TopicIds[i+1:]...)
+					break
+				}
+			}
+			// Reset active topics per block
+			activeTopicIds.TopicIds = newActiveTopicIds
+			//blockToActiveTopics[blockHeight] = types.TopicIds{TopicIds: newActiveTopicIds}
+			// Reset lowest weight per block
+			cuLowestWeight = getLowestTopicIdWeightPair(topicWeightData, blockToActiveTopics[blockHeight])
+		}
+		blockToActiveTopics[blockHeight] = activeTopicIds
+		blockHeightBytes := make([]byte, 8)
+		binary.BigEndian.PutUint64(blockHeightBytes, uint64(blockHeight))
+		churningBlockStore.Set(idArray, blockHeightBytes)
+		activeTopicsBytes, err := activeTopicIds.Marshal()
+		if err != nil {
+			continue
+		}
+		lowestWeightBytes, err := cuLowestWeight.Marshal()
+		if err != nil {
+			continue
+		}
+		blockToActiveStore.Set(blockHeightBytes, activeTopicsBytes)
+		blockLowestWeightStore.Set(blockHeightBytes, lowestWeightBytes)
+	}
+	_ = iterator.Close()
+	return nil
+}
```

Comment:
Any error in this function should panic; if an error occurs in the migration and we just `continue`, we will end up in an incorrect state.